js/switcheo.js

```diff
@@ -0,0 +1,370 @@
+                        'withdrawals/{id}/broadcast',
+                    ],
+                },
+            },
+            'exceptions': {
+                'param_required': ExchangeError, // 400 Missing parameter
+                'validation_error': ExchangeError, // 400 Unable to validate POST/PUT
+                'invalid_request': ExchangeError, // 400 Invalid request
+                'authentication_error': AuthenticationError, // 401 Invalid auth (generic)
+                'invalid_scope': AuthenticationError, // 403 User hasn’t authenticated necessary scope
+                'not_found': ExchangeError, // 404 Resource not found
+                'rate_limit_exceeded': DDoSProtection, // 429 Rate limit exceeded
+                'internal_server_error': ExchangeError, // 500 Internal server error
+            },
+            'options': {
+                'contract': '',
+            },
+        });
+    }
+
+    async fetchTime () {
+        let response = await this.publicGetExchangeTimestamp ();
+        return response['timestamp'];
+    }
+
+    parseCurrentContract (contracts) {
+        let i = 0;
+        let j = 0;
+        let switcheoContracts = contracts;
+        let current_contract = {};
+        let switcheoBlockchains = Object.keys (switcheoContracts);
+        for (i = 0; i < switcheoBlockchains.length; i++) {
+            let maxVersionNumber = 0;
+            let maxVersionKey = '';
+            let key = switcheoBlockchains[i];
+            let contractVersion = Object.keys (switcheoContracts[key]);
+            for (j = 0; j < contractVersion.length; j++) {
+                let versionNumber = parseFloat (contractVersion[j].slice (1).replace ('_', '.'));
+                if (versionNumber > maxVersionNumber) {
+                    maxVersionNumber = versionNumber;
+                    maxVersionKey = contractVersion[j];
+                }
+            }
+            current_contract[key] = switcheoContracts[key][maxVersionKey];
+        }
+        current_contract['GAS'] = current_contract['NEO'];
+        current_contract['SWTH'] = current_contract['NEO'];
+        current_contract['SDUSD'] = current_contract['NEO'];
+        current_contract['SDUSDC'] = current_contract['NEO'];
+        current_contract['PAX'] = current_contract['ETH'];
+        current_contract['DAI'] = current_contract['ETH'];
+        return current_contract;
+    }
+
+    async fetchContractHash () {
+        let response = await this.publicGetExchangeContracts ();
+        return this.parseCurrentContract (response);
+    }
+
+    async fetchMarkets (params = {}) {
+        let markets = await this.publicGetExchangePairs (this.extend ({
+            'show_details': 1,
+        }, params));
+        let tokens = await this.publicGetExchangeTokens (this.extend ({
+            'show_listing_details': 1,
+            'show_inactive': 1,
+        }, params));
+        let contracts = await this.fetchContractHash ();
```

Comment:
This looks like a lot of requests for one method. In general, the rule of thumb throughout the lib is to make as less HTTP requests as possible (ideally – not more than 1 http request per unified call). We should only do more than one call, if it absolutely required and absolutely impossible to do it otherway. In this particular case, however, I think, we should keep `publicGetExchangePairs` in `fetchMarkets`, and move `publicGetExchangeTokens` to a standalone method called `fetchCurrencies`.

Reply:
Moved `publicGetExchangeTokens` outside of the `fetchMarkets` function so that each function only has one HTTP request.  I couldn't find an expected payload in the docs from the `fetchCurrencies` function so it is only an external call.

This commit should reflect the changes: https://github.com/ccxt/ccxt/pull/4822/commits/8f59149dca9e2eb21ebdab4bd483fd93a375f217